### PR TITLE
Queries in the Misk Hibernate API.

### PR DESCRIPTION
**Pull request by swankjesse**
Currently implemented manually. Will implement with reflection.

### CODE REVIEW

File: misk/src/main/kotlin/misk/hibernate/Query.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package misk.hibernate
+
+import kotlin.reflect.KClass
+
+/** Base class for SQL queries. */
+interface Query<T> {
+  fun uniqueResult(session: Session): T?
+  fun <P : Projection> uniqueResultAs(session: Session, projection: KClass<P>): P?
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/hibernate/Query.kt]: should there be a method that does not return a nullable? I think a common pattern is just throwing a not found exception

File: misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import java.time.LocalDate
+import javax.inject.Singleton
+import javax.persistence.criteria.CompoundSelection
+import javax.persistence.criteria.CriteriaBuilder
+import javax.persistence.criteria.Path
+import javax.persistence.criteria.Predicate
+import javax.persistence.criteria.Root
+import kotlin.reflect.KClass
+
+// TODO(jwilson): build this with metaprogramming (reflection/codegen) instead of code.
```
Comment (ryanhall07) [inline misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt]: we must go deeper :)

File: misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      predicate = PredicateSpec.Lt(upperBound))
+
+  override fun uniqueResult(session: Session): DbCharacter? {
+    // TODO(jwilson): max results = 2
+    val list = list(session)
+    require(list.size <= 1) { "expected at most 1 result but was $list" }
+    return list.firstOrNull()
+  }
+
+  override fun <P : Projection> uniqueResultAs(session: Session, projection: KClass<P>): P? {
```
Comment (ryanhall07) [inline misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt]: can this just be a default impl on the interface? doesn't seem specific to the entity

File: misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt (hunk 2 above)
Comment (ryanhall07) [reply misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt]: I wonder if we could do something cool and use the indexes of the schema to generate allowable where clauses.

File: misk/src/test/kotlin/misk/hibernate/TransacterTest.kt [hunk 4]
```diff
@@ -13,23 +14,83 @@ class TransacterTest {
...
+  annotation class Constraint(
+    val path: String,
+    val operator: String = "="
+  )
+
+  annotation class Property(
+    val value: String
+  )
+
+  interface CharacterQuery : Query<DbCharacter> {
```
Comment (ryanhall07) [inline misk/src/test/kotlin/misk/hibernate/TransacterTest.kt]: should this live outside of the test?

File: misk/src/test/kotlin/misk/hibernate/TransacterTest.kt [hunk 5]
```diff
@@ -13,23 +14,83 @@ class TransacterTest {
...
+    @Constraint("movie.release_date")
```
Comment (ryanhall07) [inline misk/src/test/kotlin/misk/hibernate/TransacterTest.kt]: I don't follow how these annotations are used

Comment (swankjesse) [reply misk/src/test/kotlin/misk/hibernate/TransacterTest.kt]: when I rebuild the implementation to use reflection they will be used

File: misk/src/test/kotlin/misk/hibernate/TransacterTest.kt (hunk 4 above)
Comment (swankjesse) [reply misk/src/test/kotlin/misk/hibernate/TransacterTest.kt]: tough. The point of the test is to confirm that the query bindings work.

File: misk/src/test/kotlin/misk/hibernate/TransacterTest.kt (hunk 5 above)
Comment (ryanhall07) [reply misk/src/test/kotlin/misk/hibernate/TransacterTest.kt]: ah. spent way too long trying to figure that one out :)

File: misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt (hunk 2 above)
Comment (swankjesse) [reply misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt]: we can, but it’s an unbounded set

File: misk/src/main/kotlin/misk/hibernate/Query.kt (hunk 1 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/hibernate/Query.kt]: I’m trying to make it as easy as possible to migrate from Service Container, which returns null if there’s nothing found.

Comment (swankjesse) [reply misk/src/main/kotlin/misk/hibernate/Query.kt]: `!!` is good syntax to say “this is definitely here”

File: misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt (hunk 3 above)
Comment (swankjesse) [reply misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt]: possibly? I’m reluctant to use default methods but that’s not well motivated

File: misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt (hunk 2 above)
Comment (ryanhall07) [reply misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt]: i'm thinking we separate filter building into 2 steps, indexed and unindexed. you must pick an available index filter and then can add an additional unbounded set of unindexed filters.

we could even promote the usage of covered indices by combining indexed fields with a projection.

File: misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt (hunk 3 above)
Comment (ryanhall07) [reply misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt]: probably better to start out without them until we see the same pattern repeating over and over again

File: misk/src/main/kotlin/misk/hibernate/Query.kt (hunk 1 above)
Comment (tirsen) [reply misk/src/main/kotlin/misk/hibernate/Query.kt]: There's been plenty of bugs where we've just assumed we get something out of `uniqueResult` so I think having a reminder to think "what if this is not here?" is a good thing.

File: misk/src/main/kotlin/misk/hibernate/Query.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+  /** Creates instances of queries. */
+  interface Factory {
+    fun <T : Query<*>> newQuery(queryClass: KClass<T>): T
+  }
+}
+
+inline fun <reified P : Projection> Query<*>.listAs(session: Session) = listAs(session, P::class)
+
+inline fun <reified P : Projection> Query<*>.uniqueResultAs(session: Session) = uniqueResultAs(
```
Comment (tirsen) [inline misk/src/main/kotlin/misk/hibernate/Query.kt]: YEAH!

**Pull request merged**